crates/catalog/sql/src/catalog.rs

```diff
@@ -0,0 +1,397 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use futures::{lock::Mutex, AsyncReadExt, AsyncWriteExt};
+use sqlx::{
+    any::{install_default_drivers, AnyConnectOptions, AnyRow},
+    AnyConnection, ConnectOptions, Connection, Row,
+};
+use std::collections::HashMap;
+
+use iceberg::{
+    io::FileIO, spec::TableMetadata, table::Table, Catalog, Error, ErrorKind, Namespace,
+    NamespaceIdent, Result, TableCommit, TableCreation, TableIdent,
+};
+use uuid::Uuid;
+
+use crate::error::from_sqlx_error;
+
+#[derive(Debug)]
+/// Sql catalog implementation.
+pub struct SqlCatalog {
+    name: String,
+    connection: Arc<Mutex<AnyConnection>>,
+    storage: FileIO,
+    cache: Arc<DashMap<TableIdent, (String, TableMetadata)>>,
+}
+
+impl SqlCatalog {
+    /// Create new sql catalog instance
+    pub async fn new(url: &str, name: &str, storage: FileIO) -> Result<Self> {
+        install_default_drivers();
+
+        let mut connection = AnyConnectOptions::connect(
+            &AnyConnectOptions::from_url(&url.try_into()?).map_err(from_sqlx_error)?,
+        )
+        .await
+        .map_err(from_sqlx_error)?;
```

Comment:
Might we want to do something similar to what we have for errors in the main `iceberg` crate (eg like we do with `Reqwest` errors [here](https://github.com/apache/iceberg-rust/blob/main/crates/iceberg/src/error.rs#L310-L314)), but for `Sqlx` errors, so that we can remove all of the calls to `.map_err(from_sqlx_error)` to make it a bit cleaner?

Reply:
The problem with that is that we would need to add sqlx as a dependency for the main crate. To keep the main crate slim, I added it here.

Reply:
Aah of course, and I guess the orphan rule prevents us from extending `iceberg::Error` inside the sql catalog crate 😬